planex/extract.py

```diff
@@ -0,0 +1,184 @@
+"""
+planex-extract: Extract files from a tarball as described by a link file
+"""
+
+import argparse
+import json
+import logging
+import os
+import os.path
+import re
+import sys
+import tarfile
+import urlparse
+
+import argcomplete
+
+from planex.util import add_common_parser_options
+from planex.util import setup_logging
+from planex.util import setup_sigint_handler
+import planex.spec
+
+
+def all_sources(spec, topdir, check_package_names):
+    """
+    Get all sources defined in the spec file
+    """
+    spec = planex.spec.Spec(spec, topdir=topdir,
+                            check_package_name=check_package_names)
+    urls = [urlparse.urlparse(url) for url in spec.source_urls()]
+    return zip(spec.source_paths(), urls)
+
+
+def extract_file(tar, name_in, name_out):
+    """
+    Extract a file from a tarball
+    """
+    logging.debug("Extracting %s to %s", name_in, name_out)
+    if name_in not in tar.getnames():
+        sys.exit("%s: %s not found" % (sys.argv[0], name_in))
+    try:
+        fh_in = tar.extractfile(name_in)
+        with open(name_out, 'w') as fh_out:
+            for line in fh_in:
+                fh_out.write(line)
+    finally:
+        fh_in.close()
+
+
+def parse_patchseries(series, guard=''):
+    """
+    Parse series file and return the list of patches
+    """
+    guard_re = re.compile(r'([\S]+)(\s#.*)?')
+
+    for line in series:
+        line = line.strip()
+        if not line or line.startswith('#'):
+            continue
+
```

Comment:
That won't have the right behaviour with negative patchguards. They need to fall through to the yield().